fvm/evm/handler/blockHashList.go

```diff
@@ -0,0 +1,266 @@
+	height   uint64 // keeps the height of last added block
+}
+
+// NewBlockHashList creates a block hash list
+// It tries to load the metadata from the backend
+// and if not exist it creates one
+func NewBlockHashList(
+	backend types.Backend,
+	rootAddress flow.Address,
+	capacity int,
+) (*BlockHashList, error) {
+	bhl := &BlockHashList{
+		backend:     backend,
+		rootAddress: rootAddress,
+		capacity:    capacity,
+		tail:        0,
+		count:       0,
+		height:      0,
+	}
+	err := bhl.loadMetaData()
+	if err != nil {
+		return nil, err
+	}
+	// check the loaded capacity against the one provided
+	if bhl.capacity != capacity {
+		return nil, fmt.Errorf(
+			"capacity doesn't match, expected: %d, got: %d",
+			bhl.capacity,
+			capacity,
+		)
+	}
+	return bhl, nil
+}
+
+// Push pushes a block hash for the next height to the list.
+// If the list has reached to the capacity, it overwrites the oldest element.
+func (bhl *BlockHashList) Push(height uint64, bh gethCommon.Hash) error {
+	// handle the very first block
+	if bhl.IsEmpty() && height != 0 {
+		return fmt.Errorf("out of the order block hash, expected: 0, got: %d", height)
+	}
+	// check the block heights before pushing
+	if !bhl.IsEmpty() && height != bhl.height+1 {
+		return fmt.Errorf("out of the order block hash, expected: %d, got: %d", bhl.height+1, height)
+	}
+	// updates the block hash stored at index
+	err := bhl.updateBlockHashAt(bhl.tail, bh)
+	if err != nil {
+		return err
+	}
+
+	// update meta data
+	bhl.tail = (bhl.tail + 1) % bhl.capacity
+	bhl.height = height
+	if bhl.count != bhl.capacity {
+		bhl.count++
+	}
+	return bhl.storeMetaData()
+}
+
+// IsEmpty returns true if the list is empty
+func (bhl *BlockHashList) IsEmpty() bool {
+	return bhl.count == 0
+}
+
+// LastAddedBlockHash returns the last block hash added to the list
+// for empty list it returns empty hash value
+func (bhl *BlockHashList) LastAddedBlockHash() (gethCommon.Hash, error) {
+	if bhl.IsEmpty() {
+		// return empty hash
+		return gethCommon.Hash{}, nil
+	}
+	index := bhl.tail - 1
+	if index < 0 {
+		index = bhl.capacity - 1
+	}
+	return bhl.getBlockHashAt(index)
+}
+
+// MinAvailableHeight returns the min available height in the list
+func (bhl *BlockHashList) MinAvailableHeight() uint64 {
+	if bhl.IsEmpty() {
+		return 0
+	}
+	return bhl.height - (uint64(bhl.count) - 1)
+}
+
+// MaxAvailableHeight returns the max available height in the list
+func (bhl *BlockHashList) MaxAvailableHeight() uint64 {
+	return bhl.height
+}
+
+// BlockHashByIndex returns the block hash by block height
+func (bhl *BlockHashList) BlockHashByHeight(height uint64) (found bool, bh gethCommon.Hash, err error) {
+	if bhl.IsEmpty() ||
+		height > bhl.MaxAvailableHeight() ||
+		height < bhl.MinAvailableHeight() {
+		return false, gethCommon.Hash{}, nil
+	}
+	// calculate the index to lookup
+	diff := bhl.height - height
+	index := bhl.tail - int(diff) - 1
+	if index < 0 {
+		index = bhl.capacity + index
+	}
```

Comment:
a suggestion to avoid negative numbers altogether
```suggestion
	index := (bhl.tail - int(diff) - 1 + bhl.capacity) % bhl.capacity
```